src/algorithm/distance.rs

```diff
@@ -61,38 +60,32 @@ pub trait Distance<T, Rhs = Self> {
     /// // A Random point outside the LineString
     /// let p = Point::new(5.5, 2.1);
     /// let dist = p.distance(&ls);
-    /// assert_eq!(dist, 1.1313708498984758);
+    /// assert_eq!(dist, 1.1313708498984762);
     /// ```
     fn distance(&self, rhs: &Rhs) -> T;
 }
 
 // Return minimum distance between a Point and a Line segment
 // This is a helper for Point-to-LineString and Point-to-Polygon distance
-// adapted from http://stackoverflow.com/a/1501725/416626. Quoting the author:
-//
-// The projection of point p onto a line is the point on the line closest to p.
-// (and a perpendicular to the line at the projection will pass through p).
-// The number t is how far along the line segment from start to end that the projection falls:
-// If t is 0, the projection falls right on start; if it's 1, it falls on end; if it's 0.5,
-// then it's halfway between. If t is less than 0 or greater than 1, it
-// falls on the line past one end or the other of the segment. In that case the
-// distance to the segment will be the distance to the nearer end
+// adapted from https://github.com/OSGeo/geos/blob/master/src/algorithm/CGAlgorithms.cpp#L191
 fn line_segment_distance<T>(point: &Point<T>, start: &Point<T>, end: &Point<T>) -> T
-    where T: Float + ToPrimitive
+where
+    T: Float + ToPrimitive,
 {
-    let dist_squared = pow(start.distance(end), 2);
-    // Implies that start == end
-    if dist_squared.is_zero() {
-        return pow(point.distance(start), 2);
-    }
-    // Consider the line extending the segment, parameterized as start + t (end - start)
-    // We find the projection of the point onto the line
-    // This falls where t = [(point - start) . (end - start)] / |end - start|^2, where . is the dot product
-    // We constrain t to a 0, 1 interval to handle points outside the segment start, end
-    let t = T::zero().max(T::one().min((*point - *start).dot(&(*end - *start)) / dist_squared));
-    let projected = Point::new(start.x() + t * (end.x() - start.x()),
-                               start.y() + t * (end.y() - start.y()));
-    point.distance(&projected)
+    if start == end {
+        return point.distance(start);
+    }
+    let r = ((point.x() - start.x()) * (end.x() - start.x()) + (point.y() - start.y()) * (end.y() - start.y())) /
+        ((end.x() - start.x()) * (end.x() - start.x()) + (end.y() - start.y()) * (end.y() - start.y()));
+    if r <= T::zero() {
```

Comment:
I think this would be easier to read as a `match` statement on `r.partial_cmp(T::zero())`, but that's just my personal preference so feel free to ignore.